### ISSUE TITLE
fix fd leak in libpcp_web discovery metadata handling (on error path)

### ISSUE THREAD

**Issue by goodwinos** [performancecopilot/pcp#688]
error path occurs for a zero length data volume - early error return leaks metadata fd.

**Comment by goodwinos** [comment]
This issue is not just leaking file descriptors, but those file descriptors are holding references to deleted files (such as PCP log volumes and meta files that have been deleted because they have been compressed, thus new filename and inode).

Here's evidence of the fd leak and of the filesystem space returned  when pmproxy is restarted (releasing the open file refs)

```
shack:root@/var/log/redis[]# sudo ls -l /proc/$(pidof pmproxy)/fd
total 0
lrwx------. 1 root root 64 Nov  7 13:55 0 -> 'anon_inode:[eventpoll]'
l-wx------. 1 root root 64 Nov  7 13:55 1 -> /var/log/pcp/pmproxy/pmproxy.log
l-wx------. 1 root root 64 Nov  7 13:55 10 -> 'pipe:[122963]'
lrwx------. 1 root root 64 Nov  7 13:55 11 -> 'anon_inode:[eventfd]'
lr-x------. 1 root root 64 Nov  7 13:55 12 -> /dev/null
lrwx------. 1 root root 64 Nov  7 13:55 13 -> 'socket:[122964]'
lr-x------. 1 root root 64 Nov  7 13:55 14 -> 'pipe:[122965]'
l-wx------. 1 root root 64 Nov  7 13:55 15 -> 'pipe:[122965]'
lrwx------. 1 root root 64 Nov  7 13:55 16 -> 'socket:[122966]'
lrwx------. 1 root root 64 Nov  7 13:55 17 -> 'socket:[114538]'
lrwx------. 1 root root 64 Nov  7 13:55 18 -> 'socket:[119194]'
lrwx------. 1 root root 64 Nov  7 13:55 19 -> 'socket:[118293]'
l-wx------. 1 root root 64 Nov  7 13:55 2 -> /var/log/pcp/pmproxy/pmproxy.log
lrwx------. 1 root root 64 Nov  7 13:55 20 -> 'socket:[119197]'
lr-x------. 1 root root 64 Nov  7 13:55 21 -> anon_inode:inotify
lr-x------. 1 root root 64 Nov  7 13:55 22 -> '/var/log/pcp/pmlogger/shack/20191106.08.11.0 (deleted)'
lr-x------. 1 root root 64 Nov  7 13:55 23 -> /var/log/pcp/pmlogger/shack/20191106.08.11.index
lr-x------. 1 root root 64 Nov  7 13:55 24 -> '/var/log/pcp/pmlogger/shack/20191106.08.11.meta (deleted)'
lr-x------. 1 root root 64 Nov  7 13:55 25 -> '/var/log/pcp/pmlogger/shack/20191106.08.11.meta (deleted)'
lr-x------. 1 root root 64 Nov  7 13:55 26 -> '/var/log/pcp/pmlogger/shack/20191106.08.11.0 (deleted)'
lr-x------. 1 root root 64 Nov  7 13:55 27 -> /var/log/pcp/pmlogger/shack/20191106.08.11.index
lr-x------. 1 root root 64 Nov  7 13:55 28 -> '/var/log/pcp/pmlogger/shack/20191106.08.11.meta (deleted)'
lr-x------. 1 root root 64 Nov  7 13:55 29 -> '/var/log/pcp/pmlogger/shack/20191106.08.13.0 (deleted)'
lr-x------. 1 root root 64 Nov  7 13:55 3 -> 'pipe:[122961]'
lr-x------. 1 root root 64 Nov  7 13:55 30 -> /var/log/pcp/pmlogger/shack/20191106.08.13.index
lr-x------. 1 root root 64 Nov  7 13:55 31 -> '/var/log/pcp/pmlogger/shack/20191106.08.13.meta (deleted)'
lr-x------. 1 root root 64 Nov  7 13:55 32 -> '/var/log/pcp/pmlogger/shack/20191106.08.13.meta (deleted)'
lr-x------. 1 root root 64 Nov  7 13:55 33 -> '/var/log/pcp/pmlogger/shack/20191106.08.13.0 (deleted)'
lr-x------. 1 root root 64 Nov  7 13:55 34 -> /var/log/pcp/pmlogger/shack/20191106.08.13.index
lr-x------. 1 root root 64 Nov  7 13:55 35 -> '/var/log/pcp/pmlogger/shack/20191106.08.13.meta (deleted)'
lr-x------. 1 root root 64 Nov  7 13:55 36 -> '/var/log/pcp/pmlogger/shack/20191106.08.13.1 (deleted)'
lr-x------. 1 root root 64 Nov  7 13:55 37 -> /var/log/pcp/pmlogger/shack/20191106.08.13.index
lr-x------. 1 root root 64 Nov  7 13:55 38 -> '/var/log/pcp/pmlogger/shack/20191106.08.13.meta (deleted)'
lr-x------. 1 root root 64 Nov  7 13:55 39 -> '/var/log/pcp/pmlogger/shack/20191106.08.13.2 (deleted)'
l-wx------. 1 root root 64 Nov  7 13:55 4 -> 'pipe:[122961]'
lr-x------. 1 root root 64 Nov  7 13:55 40 -> /var/log/pcp/pmlogger/shack/20191106.08.13.index
lr-x------. 1 root root 64 Nov  7 13:55 41 -> '/var/log/pcp/pmlogger/shack/20191106.08.13.meta (deleted)'
lr-x------. 1 root root 64 Nov  7 13:55 42 -> '/var/log/pcp/pmlogger/shack/20191106.08.13.3 (deleted)'
lr-x------. 1 root root 64 Nov  7 13:55 43 -> /var/log/pcp/pmlogger/shack/20191106.08.13.index
lr-x------. 1 root root 64 Nov  7 13:55 44 -> '/var/log/pcp/pmlogger/shack/20191106.08.13.meta (deleted)'
lr-x------. 1 root root 64 Nov  7 13:55 45 -> '/var/log/pcp/pmlogger/shack/20191106.08.13.4 (deleted)'
lr-x------. 1 root root 64 Nov  7 13:55 46 -> /var/log/pcp/pmlogger/shack/20191106.08.13.index
lr-x------. 1 root root 64 Nov  7 13:55 47 -> '/var/log/pcp/pmlogger/shack/20191106.08.13.meta (deleted)'
lr-x------. 1 root root 64 Nov  7 13:55 48 -> '/var/log/pcp/pmlogger/shack/20191107.07.10.0 (deleted)'
lr-x------. 1 root root 64 Nov  7 13:55 49 -> /var/log/pcp/pmlogger/shack/20191107.07.10.index
lr-x------. 1 root root 64 Nov  7 13:55 5 -> 'pipe:[122962]'
lr-x------. 1 root root 64 Nov  7 13:55 50 -> '/var/log/pcp/pmlogger/shack/20191107.07.10.meta (deleted)'
lr-x------. 1 root root 64 Nov  7 13:55 51 -> '/var/log/pcp/pmlogger/shack/20191107.07.10.meta (deleted)'
lr-x------. 1 root root 64 Nov  7 13:55 52 -> '/var/log/pcp/pmlogger/shack/20191107.07.10.0 (deleted)'
lr-x------. 1 root root 64 Nov  7 13:55 53 -> /var/log/pcp/pmlogger/shack/20191107.07.10.index
lr-x------. 1 root root 64 Nov  7 13:55 54 -> '/var/log/pcp/pmlogger/shack/20191107.07.10.meta (deleted)'
lr-x------. 1 root root 64 Nov  7 13:55 55 -> '/var/log/pcp/pmlogger/shack/20191107.07.12.0 (deleted)'
lr-x------. 1 root root 64 Nov  7 13:55 56 -> /var/log/pcp/pmlogger/shack/20191107.07.12.index
lr-x------. 1 root root 64 Nov  7 13:55 57 -> /var/log/pcp/pmlogger/shack/20191107.07.12.meta
lr-x------. 1 root root 64 Nov  7 13:55 58 -> /var/log/pcp/pmlogger/shack/20191107.07.12.meta
lr-x------. 1 root root 64 Nov  7 13:55 59 -> '/var/log/pcp/pmlogger/shack/20191107.07.12.0 (deleted)'
l-wx------. 1 root root 64 Nov  7 13:55 6 -> 'pipe:[122962]'
lr-x------. 1 root root 64 Nov  7 13:55 60 -> /var/log/pcp/pmlogger/shack/20191107.07.12.index
lr-x------. 1 root root 64 Nov  7 13:55 61 -> /var/log/pcp/pmlogger/shack/20191107.07.12.meta
lr-x------. 1 root root 64 Nov  7 13:55 62 -> '/var/log/pcp/pmlogger/shack/20191107.07.12.1 (deleted)'
lr-x------. 1 root root 64 Nov  7 13:55 63 -> /var/log/pcp/pmlogger/shack/20191107.07.12.index
lr-x------. 1 root root 64 Nov  7 13:55 64 -> /var/log/pcp/pmlogger/shack/20191107.07.12.meta
lr-x------. 1 root root 64 Nov  7 13:55 65 -> '/var/log/pcp/pmlogger/shack/20191107.07.12.2 (deleted)'
lr-x------. 1 root root 64 Nov  7 13:55 66 -> /var/log/pcp/pmlogger/shack/20191107.07.12.index
lr-x------. 1 root root 64 Nov  7 13:55 67 -> /var/log/pcp/pmlogger/shack/20191107.07.12.meta
lr-x------. 1 root root 64 Nov  7 13:55 68 -> /var/log/pcp/pmlogger/shack/20191107.07.12.3
lr-x------. 1 root root 64 Nov  7 13:55 69 -> /var/log/pcp/pmlogger/shack/20191107.07.12.index
lrwx------. 1 root root 64 Nov  7 13:55 7 -> 'anon_inode:[eventfd]'
lr-x------. 1 root root 64 Nov  7 13:55 70 -> /var/log/pcp/pmlogger/shack/20191107.07.12.meta
lrwx------. 1 root root 64 Nov  7 13:55 8 -> 'anon_inode:[eventpoll]'
lr-x------. 1 root root 64 Nov  7 13:55 9 -> 'pipe:[122963]'
shack:root@/var/log/redis[]# sudo ls -l /proc/$(pidof pmproxy)/fd | wc -l
72
shack:root@/var/log/redis[]# ls -l /var/log/pcp/pmlogger/shack/20191107.07.12.meta
-rw-r--r--. 1 pcp pcp 11365343 Nov  7 13:55 /var/log/pcp/pmlogger/shack/20191107.07.12.meta
shack:root@/var/log/redis[]# ls -l /var/log/pcp/pmlogger/shack/20191107.07.12.1
ls: cannot access '/var/log/pcp/pmlogger/shack/20191107.07.12.1': No such file or directory
shack:root@/var/log/redis[]# df -h /var/log/pcp
Filesystem              Size  Used Avail Use% Mounted on
/dev/mapper/shack-root  221G  195G   26G  89% /
shack:root@/var/log/redis[]# systemctl restart pmproxy
shack:root@/var/log/redis[]# df -h /var/log/pcp
Filesystem              Size  Used Avail Use% Mounted on
/dev/mapper/shack-root  221G  197G   24G  90% /
shack:root@/var/log/redis[]# sudo ls -l /proc/$(pidof pmproxy)/fd | wc -l
30
```

So we not only dropped 42 leaked file descriptors, but also regained 2GB of filesystem space.

**Comment by goodwinos** [comment]
Resolved with commits 85b773befe2c0b25ef5792b56f2467af5a7caae1 and 12d5660eb2590253fa9008960c36de2a12b8dc5c
